### DISCUSSION TITLE
Allow overriding API hostnames

### DISCUSSION THREAD

**Discussion by mikebudzynski** [Azure/api-management-developer-portal#2570] [Feature request]
Allow specifying a list of custom hostnames for an API to be displayed in the developer portal instead of the API Management's domain. Sample scenario: APIM is fronted by another service.

**Comment by ErikMogensen** [comment]
Another scenario is when using a wildcard certificate in Azure API Management and some APIs may only be called using certain hosts.

**Comment by AnRei123** [comment]
Would this feature also allows to add a local host address to the list?
And would it be possible for a portal user to extend the list with other local host addresses for different ports?

**Comment by mikebudzynski** [comment]
> Would this feature also allows to add a local host address to the list? And would it be possible for a portal user to extend the list with other local host addresses for different ports?

At this moment, we don't have the feature designed yet. Thanks for the feedback, we will incorporate it once we start working on this functionality.

**Comment by patst** [comment]
Would be great to have the path customizable as well.

We use different hostnames per API fronted by an Application Gateway and redirect them to the managed gateway with path/ host rewrites.

**Comment by smokedlinq** [comment]
We want to be able to specify which of the gateway names to display, e.g. we want our sandbox hostname to be the default and not the production gateway as we only want developers in the portal to interact with the sandbox endpoints.

**Comment by shaneholder** [comment]
I'm also interested in this feature and was wondering if using something like the X-Forwarded-Host header would be a reasonable idea.  This way in the Azure App Gateway one could configure a rule to add the appropriate X-Forwarded-Host.  Also, ideally the call to the management api hostnames would not be used for generating the sampleHostname.

**Comment by mikebudzynski** [comment]
> I'm also interested in this feature and was wondering if using something like the X-Forwarded-Host header would be a reasonable idea. This way in the Azure App Gateway one could configure a rule to add the appropriate X-Forwarded-Host. Also, ideally the call to the management api hostnames would not be used for generating the sampleHostname.

@shaneholder, could you clarify how the X-Forwarded-Host header is related to displaying a custom hostname in the developer portal? Header is strictly on the API runtime side and could be used by the API gateway, while the custom hostname in the developer portal is just API's metadata.

**Comment by shaneholder** [comment]
We are hosting the developer portal behind an application gateway.  We could add something like the x-forward-host header to indicate to the dev portal the base URL we actually want to use.  This could then be used to override the base URL.  It might be slightly hacky but it's in line with how App Services work.  I realize that there are side-band calls to go get the API information and that the URL's for the API's are not actually part of the portal but this might be better than needing to configure APIM custom hostnames and certs, esp in cases where there are requirements to have a WAF in play.

**Comment by adahlinvcrs** [comment]
@mikebudzynski  would you be able to share if this being worked on currently? This feature would make the portal a lot more usable in our case.

Thanks!

**Comment by ronaldbosma** [comment]
We're in a similar situation where APIM is behind an Azure App Gateway and the gateway hostname of APIM differs from the AGW hostname. Are there any updates on this issue?
